//
// This file is part of Taffy, a mathematical programming language.
// Copyright (C) 2016-2017 Arithmagic, LLC
//
// Taffy is free software: you can redistribute it and/or modify
// it under the terms of the GNU Lesser General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// Taffy is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU Lesser General Public License for more details.
//
// You should have received a copy of the GNU Lesser General Public License
// along with this program.  If not, see <http://www.gnu.org/licenses/>.
//

const char *__compiledExceptions =    "package org.taffy.core.exception\n"
    "\n"
    "class AbortException(Exception)\n"
    "{\n"
    "    @text, @rw\n"
    "\n"
    "    (@@) newWithText: _text\n"
    "    {\n"
    "        exception = new AbortException\n"
    "        exception setText: _text\n"
    "        exception setData: _text\n"
    "        return (exception)\n"
    "    }\n"
    "}\n"
    "\n"
    "class UserGeneratedAbortSignalException(AbortException)\n"
    "{\n"
    "    @text, @rw\n"
    "\n"
    "    (@@) newWithText: _text\n"
    "    {\n"
    "        exception = new UserGeneratedAbortSignalException\n"
    "        exception setText: _text\n"
    "        exception setData: \"user generated abort exception: #[_text]\"\n"
    "        return (exception)\n"
    "    }\n"
    "}\n"
    "\n"
    "class LocalToGlobalConversionException(Exception)\n"
    "{\n"
    "    @objectName, @rw\n"
    "\n"
    "    (@@) createObjectWithName: _objectName\n"
    "    {\n"
    "        exception = new LocalToGlobalConversionException\n"
    "        exception setObjectName: _objectName\n"
    "        exception setData: \"local variable converted to global: #[_objectName]\"\n"
    "        return (exception)\n"
    "    }\n"
    "}\n"
    "\n"
    "class UnidentifiedObjectException(Exception)\n"
    "{\n"
    "    @objectName, @rw\n"
    "    @reason, @rw\n"
    "\n"
    "    (@@) createObject: _objectName\n"
    "    {\n"
    "        exception = new UnidentifiedObjectException\n"
    "        exception setObjectName: _objectName\n"
    "        exception setData: \"unidentified object: #[_objectName]\"\n"
    "        return (exception)\n"
    "    }\n"
    "\n"
    "    (@@) createObject: _objectName reason: _reason\n"
    "    {\n"
    "        exception = new UnidentifiedObjectException\n"
    "        exception setObjectName: _objectName\n"
    "        exception setReason: _reason\n"
    "        exception setData: \"unidentified object: #[_objectName] (#[_reason])\"\n"
    "        return (exception)\n"
    "    }\n"
    "}\n"
    "\n"
    "class AbstractClassInstantiationException(Exception)\n"
    "{\n"
    "    @objectName, @rw\n"
    "\n"
    "    (@@) createObject: _objectName\n"
    "    {\n"
    "        exception = new AbstractClassInstantiationException\n"
    "        exception setObjectName: _objectName\n"
    "        exception setData: (\"object #[_objectName] is abstract and cannot \"\n"
    "                            + \"be instantiated with 'new'\")\n"
    "        return (exception)\n"
    "    }\n"
    "}\n"
    "\n"
    "class ConstantRedefinitionException(Exception)\n"
    "{\n"
    "    @objectName, @rw\n"
    "\n"
    "    (@@) createObject: _objectName\n"
    "    {\n"
    "        exception = new ConstantRedefinitionException\n"
    "        exception setObjectName: _objectName\n"
    "        exception setData: (\"object #[_objectName] is constant and \"\n"
    "                            + \"cannot be redefined\")\n"
    "        return (exception)\n"
    "    }\n"
    "}\n"
    "\n"
    "class NonConstantUseOfConstantException(Exception)\n"
    "{\n"
    "    @objectName, @rw\n"
    "\n"
    "    (@@) createObject: _objectName\n"
    "    {\n"
    "        exception = new NonConstantUseOfConstantException\n"
    "        exception setObjectName: _objectName\n"
    "        exception setData: (\"a non-constant method cannot follow a constant \"\n"
    "                            + \"method for class #[_objectName]\")\n"
    "        return (exception)\n"
    "    }\n"
    "}\n"
    "\n"
    "class DivideByZeroException(Exception)\n"
    "{\n"
    "    (@@) create\n"
    "    {\n"
    "        exception = new DivideByZeroException\n"
    "        exception setData: \"divide by zero\"\n"
    "        return (exception)\n"
    "    }\n"
    "}\n"
    "\n"
    "class EmptyListException(Exception)\n"
    "{\n"
    "    (@) init\n"
    "    {\n"
    "        super init\n"
    "        self setData: \"Empty List\"\n"
    "    }\n"
    "}\n"
    "\n"
    "class FileOpenException(Exception)\n"
    "{\n"
    "    @filename, @rw\n"
    "\n"
    "    (@@) new: _filename\n"
    "    {\n"
    "        exception = new FileOpenException\n"
    "        exception setFilename: _filename\n"
    "        exception setData: \"unable to open file: #[_filename]\"\n"
    "        return (exception)\n"
    "    }\n"
    "}\n"
    "\n"
    "class FileWriteException(Exception)\n"
    "{\n"
    "    @filename, @rw\n"
    "    @error, @rw\n"
    "\n"
    "    (@@) new: _filename error: _error\n"
    "    {\n"
    "        exception = new FileWriteException\n"
    "        exception setFilename: _filename\n"
    "        exception setError: _error\n"
    "        exception setData: \"unable to write to file: #[_filename] error: #[_error]\"\n"
    "        return (exception)\n"
    "    }\n"
    "}\n"
    "\n"
    "class InvalidCastException(Exception)\n"
    "{\n"
    "    @from, @rw\n"
    "    @to, @rw\n"
    "\n"
    "    (@@) newFrom: _from to: _to\n"
    "    {\n"
    "        exception = new InvalidCastException\n"
    "        exception setFrom: _from\n"
    "        exception setTo: _to\n"
    "        exception setData: \"invalid cast from: #[_from] to: #[_to]\"\n"
    "        return (exception)\n"
    "    }\n"
    "}\n"
    "\n"
    "class InvalidComparisonResultException(Exception)\n"
    "{\n"
    "    @value, @rw\n"
    "\n"
    "    (@@) new: _value\n"
    "    {\n"
    "        exception = new InvalidComparisonResultException\n"
    "        exception setValue: _value\n"
    "        exception setData: \"invalid comparison result: #[_value]\"\n"
    "        return (exception)\n"
    "    }\n"
    "}\n"
    "\n"
    "class InvalidHashValueException(Exception)\n"
    "{\n"
    "    @value, @rw\n"
    "\n"
    "    (@@) newWithValue: _value\n"
    "    {\n"
    "        exception = new InvalidHashValueException\n"
    "        exception setValue: _value\n"
    "        exception setData: \"invalid hash value: #[_value] (a positive scalar is needed)\"\n"
    "        return (exception)\n"
    "    }\n"
    "}\n"
    "\n"
    "class InvalidIndexesException(Exception)\n"
    "{\n"
    "    @indexes, @rw\n"
    "\n"
    "    (@@) new: _indexes\n"
    "    {\n"
    "        exception = new InvalidIndexesException\n"
    "        exception setIndexes: _indexes\n"
    "        exception setData: \"invalid indexes: #[_indexes]\"\n"
    "        return (exception)\n"
    "    }\n"
    "}\n"
    "\n"
    "class IndexOutOfBoundsException(Exception)\n"
    "{\n"
    "    @index, @rw\n"
    "\n"
    "    (@@) index: _index\n"
    "    {\n"
    "        exception = new IndexOutOfBoundsException\n"
    "        exception setIndex: _index\n"
    "        exception setData: \"index out of bounds: #[_index]\"\n"
    "        return (exception)\n"
    "    }\n"
    "}\n"
    "\n"
    "class InvalidLibraryException(Exception)\n"
    "{\n"
    "    @libraryName, @rw\n"
    "\n"
    "    (@@) newWithLibraryName: _libraryName\n"
    "    {\n"
    "        exception = new InvalidLibraryException\n"
    "        exception setLibraryName: _libraryName\n"
    "        exception setData: \"invalid library: #[_libraryName]\"\n"
    "        return (exception)\n"
    "    }\n"
    "}\n"
    "\n"
    "class InvalidMarshalledDataException(Exception)\n"
    "{\n"
    "    (@@) newObject\n"
    "    {\n"
    "        exception = new InvalidMarshalledDataException\n"
    "        exception setData: \"there are no registered connectors\"\n"
    "        return (exception)\n"
    "    }\n"
    "}\n"
    "\n"
    "class InvalidMatrixSizeException(Exception)\n"
    "{\n"
    "    @wantedRows, @rw\n"
    "    @wantedColumns, @rw\n"
    "    @haveRows, @rw\n"
    "    @haveColumns, @rw\n"
    "\n"
    "    (@@) wantedRows: _wantedRows\n"
    "      wantedColumns: _wantedColumns\n"
    "           haveRows: _haveRows\n"
    "        haveColumns: _haveColumns\n"
    "    {\n"
    "        exception = new InvalidMatrixSizeException\n"
    "        exception setWantedRows: _wantedRows\n"
    "        exception setWantedColumns: _wantedColumns\n"
    "        exception setHaveRows: _haveRows\n"
    "        exception setHaveColumns: _haveColumns\n"
    "        exception setData: (\"wanted: #[_wantedRows]x#[_wantedColumns] \"\n"
    "                            + \"have: #[_haveRows]x#[_haveColumns]\")\n"
    "        return (exception)\n"
    "    }\n"
    "}\n"
    "\n"
    "class NeedVectorException(Exception)\n"
    "{\n"
    "    (@) init\n"
    "    {\n"
    "        super init\n"
    "        self setData: \"A vector is needed for this operation\"\n"
    "    }\n"
    "}\n"
    "\n"
    "class OperationOnFunctionOfNoArgumentsException(Exception)\n"
    "{\n"
    "    (@) init\n"
    "    {\n"
    "        super init\n"
    "        self setData: \"Operation on function of no arguments\"\n"
    "    }\n"
    "}\n"
    "\n"
    "class InvalidNumberArgumentsException(Exception)\n"
    "{\n"
    "    @expected, @rw\n"
    "    @given, @rw\n"
    "\n"
    "    (@@) expected: _expected given: _given\n"
    "    {\n"
    "        exception = new InvalidNumberArgumentsException\n"
    "        exception setExpected: _expected\n"
    "        exception setGiven: _given\n"
    "        exception setData: (\"invalid number arguments, expected: #[_expected] \"\n"
    "                            + \"given: #[_given]\")\n"
    "        return (exception)\n"
    "    }\n"
    "}\n"
    "\n"
    "class InvalidReturnValueException(Exception)\n"
    "{\n"
    "    @value, @rw\n"
    "\n"
    "    (@@) newObject: _value\n"
    "    {\n"
    "        exception = new InvalidReturnValueException\n"
    "        exception setValue: _value\n"
    "        exception setData: \"invalid return value: #[_value]\"\n"
    "        return (exception)\n"
    "    }\n"
    "}\n"
    "\n"
    "class NeedByteException(Exception)\n"
    "{\n"
    "    @value, @rw\n"
    "\n"
    "    (@@) newObject: _value\n"
    "    {\n"
    "        exception = new NeedByteException\n"
    "        exception setValue: _value\n"
    "        exception setData: (\"a byte (0-255) is needed but was given: #[_value]\"\n"
    "                            + \" of type: #[_value className]\")\n"
    "        return (exception)\n"
    "    }\n"
    "}\n"
    "\n"
    "class NeedIntegerException(Exception)\n"
    "{\n"
    "    @value, @rw\n"
    "\n"
    "    (@@) newObject: _value\n"
    "    {\n"
    "        exception = new NeedIntegerException\n"
    "        exception setValue: _value\n"
    "        exception setData: (\"an integer is needed but was given: #[_value]\"\n"
    "                            + \" of type: #[_value className]\")\n"
    "        return (exception)\n"
    "    }\n"
    "}\n"
    "\n"
    "class NeedDoubleException(Exception)\n"
    "{\n"
    "    @value, @rw\n"
    "\n"
    "    (@@) newObject: _value\n"
    "    {\n"
    "        exception = new NeedDoubleException\n"
    "        exception setValue: _value\n"
    "        exception setData: (\"a double is needed but was given: #[_value]\"\n"
    "                            + \" of type: #[_value className]\")\n"
    "        return (exception)\n"
    "    }\n"
    "}\n"
    "\n"
    "class NeedPositiveIntegerException(Exception)\n"
    "{\n"
    "    @value, @rw\n"
    "\n"
    "    (@@) newObject: _value\n"
    "    {\n"
    "        exception = new NeedPositiveIntegerException\n"
    "        exception setValue: _value\n"
    "        exception setData: (\"a positive integer is needed but was \"\n"
    "                            + \"given: #[_value] of type: #[_value className]\")\n"
    "        return (exception)\n"
    "    }\n"
    "}\n"
    "\n"
    "class NoRegisteredConnectorsException(Exception)\n"
    "{\n"
    "    (@@) newObject\n"
    "    {\n"
    "        exception = new NoRegisteredConnectorsException\n"
    "        exception setData: \"there are no registered connectors\"\n"
    "        return (exception)\n"
    "    }\n"
    "}\n"
    "\n"
    "class MoreThanOnePackageException(Exception)\n"
    "{\n"
    "    @packageName, @rw\n"
    "\n"
    "    (@@) newObject: _packageName\n"
    "    {\n"
    "        exception = new MoreThanOnePackageException\n"
    "        exception setPackageName: _packageName\n"
    "        exception setData: (\"more than one package was \"\n"
    "                            + \"defined at: #[_packageName]\")\n"
    "        return (exception)\n"
    "    }\n"
    "}\n"
    "\n"
    "class PrecisionTooSmallException(Exception)\n"
    "{\n"
    "    @precision, @rw\n"
    "\n"
    "    (@@) newObject: _value\n"
    "    {\n"
    "        exception = new PrecisionTooSmallException\n"
    "        exception setPrecision: _value\n"
    "        exception setData: \"given precision is too small: #[_value]\"\n"
    "        return (exception)\n"
    "    }\n"
    "}\n"
    "\n"
    "class StackOverflowException(Exception)\n"
    "{\n"
    "    (@) init\n"
    "    {\n"
    "        super init\n"
    "        self setData: \"stack overflow\"\n"
    "    }\n"
    "}\n"
    "\n"
    "class SocketSendFailureException(Exception)\n"
    "{\n"
    "    (@@) create\n"
    "    {\n"
    "        exception = new SocketSendFailureException\n"
    "        exception setData: \"send failure\"\n"
    "        return (exception)\n"
    "    }\n"
    "}\n"
    "\n"
    "class UnidentifiedMethodException(Exception)\n"
    "{\n"
    "    @exceptionMethodName, @rw\n"
    "    @exceptionClassName, @rw\n"
    "\n"
    "    (@@) methodName: _methodName className: _className\n"
    "    {\n"
    "        exception = new UnidentifiedMethodException\n"
    "        exception setExceptionMethodName: _methodName\n"
    "        exception setExceptionClassName: _className\n"
    "        exception setData: (\"unidentified method: #[_methodName] for \"\n"
    "                            + \"class: #[_className]\")\n"
    "        return (exception)\n"
    "    }\n"
    "}\n"
    "\n"
    "class UnmarshallFailureException(Exception)\n"
    "{\n"
    "    @value, @rw\n"
    "\n"
    "    (@@) newObject: _value\n"
    "    {\n"
    "        exception = new UnmarshallFailureException\n"
    "        exception setValue: _value\n"
    "        exception setData: \"unmarshall failure for: #[_value]\"\n"
    "        return (exception)\n"
    "    }\n"
    "}\n"
    "\n"
    "class CInputException(Exception)\n"
    "{\n"
    "    (@) init\n"
    "    {\n"
    "        super init\n"
    "        self setData: \"C Input Exception\"\n"
    "    }\n"
    "}\n"
    "\n"
    "class DerivationException(Exception)\n"
    "{\n"
    "    (@) init\n"
    "    {\n"
    "        super init\n"
    "        self setData: \"invalid derivation operation\"\n"
    "    }\n"
    "}\n"
    "\n"
    "class NeedSliceException(Exception)\n"
    "{\n"
    "    (@) init\n"
    "    {\n"
    "        super init\n"
    "        self setData: \"a slice is needed for this operation\"\n"
    "    }\n"
    "}\n"
    "\n"
    "class FinalClassUpdateException(Exception)\n"
    "{\n"
    "    (@) init\n"
    "    {\n"
    "        super init\n"
    "        self setData: \"a class deemed final may not be updated\"\n"
    "    }\n"
    "}\n"
    "\n"
    "class InvalidSuperClassException(Exception)\n"
    "{\n"
    "    @superName, @rw\n"
    "\n"
    "    (@@) newWithSuperName: _name\n"
    "    {\n"
    "        exception = new InvalidSuperClassException\n"
    "        exception setSuperName: _name\n"
    "        exception setData: \"class #[_name] is final and cannot be extended\"\n"
    "        return (exception)\n"
    "    }\n"
    "}\n"
    "\n"
    "class InconsistentClassUpdateException(Exception)\n"
    "{\n"
    "    (@) init\n"
    "    {\n"
    "        super init\n"
    "        self setData: (\"A class update must keep its flags. \"\n"
    "                       + \"For example, an abstract class can only be updated \"\n"
    "                       + \"by an abstract class, and a slice can only be \"\n"
    "                       + \"updated by a slice.\")\n"
    "    }\n"
    "}\n"
    "\n"
    "class ImportFailedException(Exception)\n"
    "{\n"
    "    @packageName, @rw\n"
    "\n"
    "    (@@) newWithPackageName: _packageName\n"
    "    {\n"
    "        result = new ImportFailedException\n"
    "        result setPackageName: _packageName\n"
    "        result setData: \"Cannot find package named: #[_packageName]\"\n"
    "        return (result)\n"
    "    }\n"
    "}\n"
    "\n"
    "class ReturnWithNoCallStackException(Exception)\n"
    "{\n"
    "    (@) init\n"
    "    {\n"
    "        super init\n"
    "        self setData: (\"this return exists with no call stack\")\n"
    "    }\n"
    "}\n"
    "\n"
    "class BreakWithoutALoopException(Exception)\n"
    "{\n"
    "    (@) init\n"
    "    {\n"
    "        super init\n"
    "        self setData: (\"this break exists without a loop\")\n"
    "    }\n"
    "}\n"
    "\n"
    "class NeedMetaClassException(Exception)\n"
    "{\n"
    "    (@) init\n"
    "    {\n"
    "        super init\n"
    "        self setData: (\"a meta class is needed\")\n"
    "    }\n"
    "}\n"
    "\n"
    "class UnidentifiedClassException(Exception)\n"
    "{\n"
    "    @className, @rw\n"
    "\n"
    "    (@@) newWithName: _name\n"
    "    {\n"
    "        result = new UnidentifiedClassException\n"
    "        result setClassName: _name\n"
    "        result setData: \"Can't find class with name #[_name]\"\n"
    "        return (result)\n"
    "    }\n"
    "}\n"
    "\n"
    "class PluginInstantiationException(Exception)\n"
    "{\n"
    "    @reason, @rw\n"
    "\n"
    "    (@@) newWithReason: _reason\n"
    "    {\n"
    "        result = new PluginInstantiationException\n"
    "        result setReason: _reason\n"
    "        result setData: \"Unable to instantiate plugin: #[_reason]\"\n"
    "        return (result)\n"
    "    }\n"
    "}\n"
    "\n"
    "class SingletonInstantiationException(Exception)\n"
    "{\n"
    "    @className, @rw\n"
    "\n"
    "    (@@) newWithClassName: _className\n"
    "    {\n"
    "        result = new SingletonInstantiationException\n"
    "        result setClassName: _className\n"
    "        result setData: \"Cannot instantiate singleton class: #[_className]\"\n"
    "        return (result)\n"
    "    }\n"
    "}\n"
    "\n"
    "class UnsupportedMathOperationException(Exception)\n"
    "{\n"
    "    @reason, @rw\n"
    "\n"
    "    (@@) newWithReason: _reason\n"
    "    {\n"
    "        result = new UnsupportedMathOperationException\n"
    "        result setReason: _reason\n"
    "        result setData: \"Unsupported math operation: #[_reason]\"\n"
    "        return (result)\n"
    "    }\n"
    "}\n"
    "\n"
    "class FileOperationFailedException(Exception)\n"
    "{\n"
    "    @name, @rw\n"
    "    @reason, @rw\n"
    "\n"
    "    (@@) newWithName: _name reason: _reason\n"
    "    {\n"
    "        exception = new FileOperationFailedException\n"
    "        exception setName: _name\n"
    "        exception setReason: _reason\n"
    "        exception setData: \"file operation for '#[_name]' failed: #[_reason]\"\n"
    "        return (exception)\n"
    "    }\n"
    "}\n"
    "\n"
    "class InvalidFileModeException(Exception)\n"
    "{\n"
    "    (@) init\n"
    "    {\n"
    "        self setData: \"invalid file mode, expecting write but found read\"\n"
    "    }\n"
    "}\n"
    "\n"
    "class DeadlockException(Exception)\n"
    "{\n"
    "    (@) init\n"
    "    {\n"
    "        self setData: \"deadlock detected\"\n"
    "    }\n"
    "}\n"
    "\n"
    "class BlockNotSingularException(Exception)\n"
    "{\n"
    "    (@) init\n"
    "    {\n"
    "        self setData: \"given block is not singular\"\n"
    "    }\n"
    "}\n"
    "\n"
    "class AssertFailedException(Exception)\n"
    "{\n"
    "    (@) init\n"
    "    {\n"
    "        self setData: \"an assert failed\"\n"
    "    }\n"
    "}\n"
    "\n"
    "class InvalidFormattedNumberException(Exception)\n"
    "{\n"
    "    @numberString, @rw\n"
    "\n"
    "    (@@) newWithNumber: _number\n"
    "                 style: _style\n"
    "    {\n"
    "        exception = new InvalidFormattedNumberException\n"
    "        exception setNumberString: _number\n"
    "        exception setData: \"Invalid formatted number: #[_number] for style: #[_style]\"\n"
    "        return (exception)\n"
    "    }\n"
    "}\n"
;